stl/inc/algorithm

```diff
@@ -2096,6 +2096,53 @@ _NODISCARD _FwdIt1 find_first_of(_ExPo&& _Exec, const _FwdIt1 _First1, const _Fw
 }
 #endif // _HAS_CXX17
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::find_first_of
+    class _Find_first_of_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <input_iterator _It1, sentinel_for<_It1> _Se1, forward_iterator _It2, sentinel_for<_It2> _Se2,
+            class _Pr = ranges::equal_to, class _Pj1 = identity, class _Pj2 = identity>
+            requires indirectly_comparable<_It1, _It2, _Pr, _Pj1, _Pj2>
+        _NODISCARD constexpr _It1 operator()(_It1 _First1, const _Se1 _Last1, const _It2 _First2, const _Se2 _Last2,
+            _Pr _Pred = {}, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            _Adl_verify_range(_First1, _Last1);
+            _Adl_verify_range(_First2, _Last2);
+            auto _UFirst1       = _Get_unwrapped(_STD move(_First1));

```

Comment:
More of a question than a comment, eh?

The most fundamental C++20 iterator concepts require `movable`, but not `copyable`; `forward_iterator` adds the `copyable` requirement. In effect, types that satisfy the input and output iterator concepts could be move-only whereas forward/bidi/random access/contiguous iterators can always be copied.

Move-only iterators presumably have [some magic inside that makes them work](https://tvtropes.org/pmwiki/pmwiki.php/Main/AppliedPhlebotinum) that cannot be copied, but which move operations can relocate from one iterator into another. To work with such iterators, our iterator unwrapping machinery (`_Get_unwrapped`) must also yoink the magic out of a wrapped iterator and put it into the corresponding unwrapped iterator, to be returned later by the `_Seek_wrapped` operation.

Practically that means you'll see Range algorithms moving single-pass iterators (since they must) and copying forward-and-stronger iterators because I haven't yet completely convinced myself we should always move. Obviously moving is never more expensive than copying, I'm mostly afraid of the throughput cost of instantiated more copies of `std::move` and the readability hit of spamming `_STD move(meow)` everywhere.

I'm somewhat inclined to make what "old" algorithms I can accept "new" iterators, so I may eventually make a pass over everything `<algorithm>` and uniformly move instead of copying when possible.

Reply:
Gotcha - this helps a lot.  Thank you!